Restructure version control chapter and add subchapter on VCS for data (#859)
alan-turing-institute/the-turing-way: book/content/version_control/vcs-github.md

## Using Git commands to work on GitHub

As research becomes increasingly collaborative and multiple people work on the same project, it becomes difficult to keep track of changes made by others if not done systematically.
Moreover, it is time-consuming to manually incorporate the work of different participants in a project, even when all of their changes are compatible.

> Hosting the project on an online distributed version control system like GitHub is beneficial to make collaborations open and effective.

Author: Sorry for this nitpick. Git is a distributed version control system, GitHub is a web-based hosting service for Git repositories. Its hard to disentangle the two some times, but it is so helpful to clearly distinguish them - else people may try the GitHub documentation for problems they are having with Git commands and should consult Git's documentation for.

```suggestion
Hosting the project on an online repository hosting service like GitHub is beneficial to make collaborations open and effective.
```

Reviewer: Absolutely agree on that, and I appreciate you pointing these out.

If you are new to collaboration through [GitHub](https://github.com), please follow a comprehensive guide in the collaboration section.

In this section we will discuss how to use Git commands to work on GitHub repository.

### Create a local copy of an online repository

> There are some GitHub features that can be accessed via command-line Git tool.

Author: Just as a comment: In principle, none of these commands are specific to GitHub. Cloning, Git remotes, pushing, pulling is part of the standard Git tool kit, and were originally (and are still) used for collaborative work on Git repositories. Therefore, all of these commands are not specifically created for interactions with GitHub, but rather with interactions on any repository hosting site/server (which can be GitHub, but also GitLab, Bitbucket, Gin, or a self-set-up bare Git repository on a web server). I do not think, however, that all of this background is useful to add here at this point of the turing way and rather distracts from the main point in this chapter - which is: how do I use it efficiently. But I'm adding this comment just in case and to keep in mind for more advanced chapters.

Reviewer: That's a good point @adswa. I think we should add a note here with details you have listed in your comment. We can say that for simplicity purpose, we will use GitHub as a platform to explain these commands. Do you think it will help clarify?

> There are some GitHub features that can be accessed via command-line Git tool.

Author: Maybe drop the term "remote" here
```suggestion
So far, all Git commands introduced in this chapter concerned local, unconnected Git repositories.
In order to collaborate with others, hosting services such as GitHub can store a *clone* (i.e., copy) of your local repository and expose it to others. 
Usually, you will have a local repository and a *remote*, web-hosted repository. 
Your local repository is connected to the web-based clone. 
In technical terms, the web-based clone is a `remote` of the local repository. Usually, this remote is called "origin".
Having a web-based remote allows you to *push* changes of your project online, allows others to obtain their own clone of your repository (i.e., copy your repository to their local computer), make changes, and submit a *pull request* that allows you to integrate their changes.
```

For example, one can create an independent local copy of the project using the following Git command:

```
git clone <insert GitHub link of the repository here>
```

Collaborators can update their local version of an online repository or *pull* other's work into their own copy using the command:
```
git pull
```
Similarly, they can edit files locally and stage their updates (`git add .`), commit changes to a new version (`git commit`) and *push* changes to the remote online repository using the Git command:
```
git push
```

### Link a local project on your computer to an online repository

To link a project on your computer to a new GitHub repository (preferably with the same name) you need to follow the standard workflow for creating a Git repository (described in the [version control chapter](/version_control/01/vcs_workflow)) by using the following set of commands in the terminal one by one:

```
cd <your project folder>
git init
git add .
git commit
```

> Assuming that you have a GitHub repository that you want want to commect with this project, run the following command

Author: ```suggestion
Assuming that you have a GitHub repository that you want want to connect with this project, run the following command
```

```
git remote add origin <GitHub repository link for your project>
```

Then, *push* all the files on your computer to the online version so they match via:

```
git push -u origin master
```

> You can the go on and make more commits on your computer.

Author: ```suggestion
You can then go on and make more commits on your computer.
```

When you want to push them to your online version similarly you do:

```
git push origin branch_you_want_to_push_to
```

You can also make changes directly on the GitHub by editing the online repository, and *pull* those changes locally by using the Git commans `git pull`

Others can then clone the repository to their computer by using:

```
git clone https://GitHub.com/your_username/repository_name.Git
```

They can make and commit changes to the code without impacting the original, and push their changes to *their* online GitHub account using:

```
git push -u origin master
```

The exact same procedure applies to you if you want to clone someone else's repository.

#### Pull requests

If you are working on a personal branch and meanwhile if some other changes were made in the master branch, you can *pull* those changes down to your branch using the Git command:
```
git pull origin master
```

> When everyone's has a copy of the project on their personal branch (checkout to your branch with `git checkout branch-name`), they can *push* their changes to their branch by using the following command:

Author: ```suggestion
When everyone has a copy of the project on their personal branch (checkout your branch with `git checkout branch-name`), they can *push* their changes to their branch by using the following command:
```

Once your fork is in sync with upstream master repository, you can always keep your local cloned repository in sync with origin (fork in this case) by using:

```
git checkout master
git pull
```

The git pull command combines two other commands, git fetch and git merge.
When using fetch, the resulting commits are stored as the remote branch allows you to review the changes before merging.

Similarly, if you have created more branches other than master, you can also keep them in sync with your master, once it is in sync with the upstream repository.

```
git checkout my-other-branch
git pull origin master
```

When everything is  up-to-date, you can work on your branch and commit changes.

When you are ready to push your local commits to your forked repository (origin), use the following command.

```
git push origin forked_repository
```

Now you can make a pull request!

#### Good practice

Before you create a branch, make sure you have all the upstream changes from the origin/master branch.

> Word of caution on `rebase` command: While trying to keep your branches in sync, you may come across `rebase` command.

Author: ```suggestion
A word of caution on the `rebase` command: While trying to keep your branches in sync, you may come across `rebase` command.
```

> It tends to rewrite history and could be troublesome if not communicated with others working on the same branch, therefore try to avoid using `rebase` command, and instead use `pull` or `fetch`+`merge` as discussed in this sectoon.

Author: ```suggestion
It tends to rewrite history and could be troublesome if not communicated with others working on the same branch, therefore try to avoid using `rebase` command, and instead use `pull` or `fetch`+`merge` as discussed in this section.
```
